rcl/src/rcl/arguments.c

```diff
@@ -139,11 +143,45 @@ rcl_arguments_get_param_overrides(
   return RCL_RET_OK;
 }
 
+rcl_ret_t
+rcl_arguments_get_log_levels(
+  const rcl_arguments_t * arguments,
+  rcl_log_levels_t ** log_levels)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(arguments, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(arguments->impl, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(arguments->impl->log_levels, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(log_levels, RCL_RET_INVALID_ARGUMENT);
+
+  if (NULL != *log_levels) {
+    RCL_SET_ERROR_MSG("Output log level pointer is not null.");
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  *log_levels = rcl_log_levels_init(
+    arguments->impl->allocator, arguments->impl->log_levels->num_logger_settings);
+  if (NULL == *log_levels) {
+    return RCL_RET_BAD_ALLOC;
+  }
+  rcl_ret_t ret = rcl_log_levels_copy(arguments->impl->log_levels, *log_levels);
+  if (RCL_RET_OK != ret) {
+    rcl_ret_t log_levels_ret = rcl_log_levels_fini(*log_levels);
+    if (RCL_RET_OK != log_levels_ret) {
+      ret = log_levels_ret;
+      RCL_SET_ERROR_MSG("Error while finalizing log levels due to another error");
+    }
+    *log_levels = NULL;
+    return ret;
+  }
```

Comment:
this is wrong.

you should never need to call a fini function after a init/copy function failed.
init/copy functions should take care of cleaning up resources on failure.

note: there might be other places where we do this, but it's not the correct thing to do.